src/main/java/org/optaplanner/springboottaskassigning/solver/DefaultSolverManager.java

```diff
@@ -53,12 +52,30 @@ public DefaultSolverManager() {
     }
 
     @Override
-    public synchronized void solve(Object problemId, Solution_ planningProblem,
-                                   Consumer<Solution_> onBestSolutionChangedEvent, Consumer<Solution_> onSolvingEnded) {
-        if (problemIdToSolverTaskMap.containsKey(problemId)) {
-            throw new IllegalArgumentException("Problem (" + problemId + ") already exists.");
+    public void solve(Object problemId,
+                      Solution_ planningProblem,
+                      Consumer<Solution_> onBestSolutionChangedEvent,
+                      Consumer<Solution_> onSolvingEnded) {
+        solve(problemId, planningProblem, onBestSolutionChangedEvent, onSolvingEnded, null);
+    }
+
+    @Override
+    public void solve(Object problemId,
+                      Solution_ planningProblem,
+                      Consumer<Solution_> onBestSolutionChangedEvent,
+                      Consumer<Solution_> onSolvingEnded,
+                      Consumer<Throwable> onException) {
+        SolverTask<Solution_> newSolverTask;
+        synchronized (this) {
```

Comment:
Nice catch!
I don't think the second approach is valid, since the problem might be submitted by the time Thread B is in the if block before logging the error message, so it will log an incorrect error message.
These methods should use synchronization, and it won't be a problem since they're not called very frequently and the locking is only for checking & retrieving solverTasks from the map so the effect on performance is negligible.

Reply:
See my latest comment - I think there is more elegant solution.